Faces branch on top of master

ricardofantin opened the pull request:
All changes from mine faces branch are included in this single commit.

One of the main problems with the old branch, is the unstable commits that doesn't compile. This single commit changes everything in one step jumping over that problem.

compile with:
meson build/ -Dface=true

ricardofantin committed:
A clean-diff of faces branch on top of master

phako commented on meson.build:
```diff
@@ -81,6 +81,11 @@ if get_option('disable-dupe-detection')
   add_global_arguments(['--define=NO_DUPE_DETECTION'], language : vala)
 endif
 
+if get_option('face')
```
I would suggest something like "faces" or "face-detection"

phako commented on facedetect/shotwell-facedetect.cpp:
```diff
@@ -0,0 +1,125 @@
+#include "opencv2/objdetect/objdetect.hpp"
```
Please add the Copyright header and the original author to this.

phako commented on src/db/FaceLocationTable.vala:
```diff
@@ -0,0 +1,202 @@
+/* Copyright 2011-2012 Yorba Foundation
```
Please update according to current copyright

phako commented on src/editing_tools/EditingTools.vala:
```diff
@@ -1815,6 +1815,958 @@ public class CropTool : EditingTool {
+        DETECTING_FACES_FINISHED
+    }
+
+    public class FaceWidget : Gtk.HBox {
```
Please fix deprecated Gtk+ widgets such as HBox => Box with orientation: horizontal

phako commented on src/editing_tools/EditingTools.vala:
```diff
@@ -1815,6 +1815,958 @@ public class CropTool : EditingTool {
+        public FaceWidget (FaceShape face_shape) {
+            spacing = CONTROL_SPACING;
+
+            edit_button = new Gtk.Button.from_stock(Gtk.Stock.EDIT);
```
Same for stock buttons, those are deprecated as well

phako commented on src/editing_tools/EditingTools.vala:
```diff
@@ -1815,6 +1815,958 @@ public class CropTool : EditingTool {
+                assert_not_reached();
+            }
+            
+            faces = new Gee.PriorityQueue<string>();
```
As a backlog item, we should move this to something that's slightly nicer to parse than this

phako commented on src/editing_tools/EditingTools.vala:
```diff
@@ -1815,6 +1815,958 @@ public class CropTool : EditingTool {
 
 }
 
+#if ENABLE_FACES
```
I think this is big enough to split it off into its own file

phako commented on src/faces/Face.vala:
```diff
@@ -0,0 +1,682 @@
+    
+    protected override void notify_container_contents_removed(ContainerSource container, 
+        Gee.Collection<DataObject> removed, bool unlinking) {
+//        Gee.Collection<DataSource> removed, bool unlinking) {
```
No commented code - why is that here?

phako commented on src/faces/FacePage.vala:
```diff
@@ -0,0 +1,139 @@
+
+    private const GLib.ActionEntry[] entries = {
+        { "DeleteFace", on_delete_face },
+		{ "RenameFace", on_rename_face },
```
Indentation error

phako commented on src/faces/FacePage.vala:
```diff
@@ -0,0 +1,139 @@
+        
+        //groups.add_menu_item(_("Face Location"), "FaceLocation");
+       
+        //groups += create_photos_faces_injectables();
```
Commented code

phako commented on src/faces/FacePage.vala:
```diff
@@ -0,0 +1,139 @@
+       
+        //groups += create_photos_faces_injectables();
+		groups += create_faces_menu_injectables();
+        
```
Indendation error

phako commented on src/faces/FacePage.vala:
```diff
@@ -0,0 +1,139 @@
+        //menuFaces.add_menu("Faces");
+        menuFaces.add_menu_item("Remove Face \"" + this.face.get_name() + "\" From Photos", "RemoveFaceFromPhotos", "<Primary>r");
+        menuFaces.add_menu_item("Rename Face \"" + this.face.get_name() + "\"…", "RenameFace", "<Primary>e");
+        menuFaces.add_menu_item("Dele_teFace", "DeleteFace", "<Primary>t");
```
Missing white space, also texts should be translatable

phako commented on src/library/LibraryWindow.vala:
```diff
@@ -269,6 +279,148 @@ public class LibraryWindow : AppWindow {
     }
 
+/*<<<<<<< HEAD
+        Gtk.ActionEntry preferences = { "CommonPreferences", Resources.PREFERENCES_LABEL, TRANSLATABLE,
```
Do you really need to keep this all in here?

phako commented on src/meson.build:
```diff
@@ -17,13 +17,24 @@ processor = executable('shotwell-graphics-processor',
                        dependencies: [gio, gdk, gee],
                        link_with: sw_graphics_processor)
 
+face_resources = []
```
Technically, that's face_sources; resources is used for the resource files (ui, icons etc.)

phako reviewed (changes requested):
Some things need to be cleaned up. After that, I think we can merge it (and after the feature freeze). But the faces stuff definitely needs some more work after that

ricardofantin commented:
Thank you for the comments. Saturday I should make all changes.

In my next changes I will take more care about tabs and comments.

ricardofantin commented:
Hello.
I made almost everything you asked, the only thing left is to correctly write the code to replace the deprecated stock buttons. I also noted that are many places with code to use the faces. To make the code better I should reduce the places, besides the src/faces/ folder, which have faces related code. In some files there are 4 different places with faces related code. I will try to reduce it. After removing the code from src/editing_tools/EditingTools.vala and creating src/faces/FacesTool.vala there are more 13 files with faces related code. It will take a while to get the code better.
The "-Dface=true" were taken from the Makefile from the old faces code. I chose to use "-Dface-detection=true".
I have many questions, if you could answer some of them would really help me:
1. Which IDE you use to edit vala? (I'm used vim and devhelp, and now I'm using geany)
2. To make a text translatable, I can make an entry in the ui/faces.ui or put the string in src/Resource.vala, right? Is there another way? I'm almost sure that the better solution is make the entry in ui/faces.ui, is it right?
3. Instead of using the Gtk.Button.from_stock(Gtk.Stock.EDIT); I'm creating a new Button and taking the text from Resources.EDIT_LABEL, but the button appear with label "_Edit" . There should not be the underline before the Edit. How do I remove it and make the system use the translated text? Another thing we should discuss is the icons, some icons from the stock buttons are not available through icon name. So which icons I should use in the buttons? The new icons and the text with underlines are attached. The old icons used are attached too, the text in this version is in Brazilian Portuguese, but only look at the icons.
![old-icons-from-stockbutton](https://user-images.githubusercontent.com/5272036/37127798-1ec8cc76-2256-11e8-8a05-a96e17523f37.png)
![_edit-and-icones](https://user-images.githubusercontent.com/5272036/37127481-c19075f0-2254-11e8-8dd5-589383492672.png)
4. In the copyright header, can I put the year of 2018?
5. I made the facedetect/shotwell-facedetect.cpp copyright header this way:

```
/* Copyright 2016 Software Freedom Conservancy Inc.
 * 
 * Copyright 2011 Valentín Barros Puertas <valentin(at)sanva(dot)net>
 * Copyright 2018 Ricardo Fantin da Costa <ricardofantin(at)gmail(dot)com>
 * 
 * This software is licensed under the GNU LGPL (version 2.1 or later).
 * See the COPYING file in this distribution.
 */
```
Is this copyright ok?

Thank you for your time. I'll try to make the better code I can.

phako commented:
1. I use vim or gnome-builder occasionally
2. Yes, please put it in the ui file where possible. If you use strings in vala code, plesae make sure to prefix it with _, e.g. _("Some text to translate")
3. In Glade, check the "Always use underline" checkbox. In code, use new Button.with_mnemonic() or set use_underline to true
4. Yes, use 2018. The header also looks fine

phako commented:
1. I use vim or gnome-builder occasionally
2. Yes, please put it in the ui file where possible. If you use strings in vala code, plesae make sure to prefix it with _, e.g. _("Some text to translate")
3. In Glade, check the "Always use underline" checkbox. In code, use new Button.with_mnemonic() or set use_underline to true. For the icon use "gtk-edit" - that's basically cheating, but I don't know any better solution for now
4. Yes, use 2018. The header also looks fine

ricardofantin committed:
Faces code clean up

Change the meson option from face to face-detection.
To compile with face support use -Dface-detection=true.
Many files used old Yorba Copyright notice, they were updated.
A class from file src/editing_tools/EditingTools.vala was moved to it own file
src/faces/FacesTools.vala .
A bug made the menu label to be always in plural "Remove .. photos" instead of
"Remove ... photo". Now it is corrected.
The deprecateds Gtk.HBox and Gtk.Button.from_stock() where changed to they
correct class.

ricardofantin commented:
Hello.
All requested changes were made.
The commit message resumes all major changes. Some minor changes include substitution uncorrected tabs used in indentation to spaces, remove of many commented code.

I start make the Makefile.ac in an old commit (Not included is this push). I want to finish this code to the use can compile in two ways. And make a name autocompletion to finish the bug [717733](https://bugzilla.gnome.org/show_bug.cgi?id=717733) .

Some things to make in the future:
In src/faces/FacesTool.vala:
faces = new Gee.PriorityQueue<string>();
"As a backlog item, we should move this to something that's slightly nicer to parse than this"

In many files there are Faces related code in different places, maybe put everything just in one place.

Any feedback, is welcome.

Bye.

phako commented:
Hi, sorry, i was busy. Please forget about Makefile.ac. It is gone on master. I will do a deeper check tomorrow

phako commented:
You forgot to add src/faces/FacesTool.vala to the commit

ricardofantin commented:
I'll make a force amend to include src/faces/FacesTool.vala and will update the meson.build version too.
Just a minute.

ricardofantin commented:
Everything correct now.

phako commented on src/faces/FacesTool.vala:
```diff
@@ -0,0 +1,971 @@
+
+                    break;
+                case EditingPhase.DETECTING_FACES_FINISHED:
+                    help_text.set_text(_("If you don't set the name of unknown faces they won't be saved."));
```
use unicode apostrophe ’ inste’ad of normal '

phako commented on src/faces/FacesTool.vala:
```diff
@@ -0,0 +1,971 @@
+                Process.spawn_sync(null, argv, null, SpawnFlags.STDERR_TO_DEV_NULL, null, out output);
+
+            } catch (SpawnError e) {
+                stderr.printf("Error trying to spawn face detection program: %s\n", e.message);
```
Make that a criticla() log message. Since this is a background job, save the error, then in the finished callback, check for the error and show it to the user with AppWindow.error()

phako commented on src/faces/FacesTool.vala:
```diff
@@ -0,0 +1,971 @@
+
+                string[] type_and_serialized = line.split(";");
+                if (type_and_serialized.length != 2) {
+                    stderr.printf("Wrong serialized line in face detection program output.");
```
use critical() instead of stderr.printf. Assert can stay, though

phako commented on src/faces/FacesTool.vala:
```diff
@@ -0,0 +1,971 @@
+                        break;
+
+                    case "warning":
+                        stderr.printf("%s\n", type_and_serialized[1]);
```
warning()

phako commented on src/faces/FacesTool.vala:
```diff
@@ -0,0 +1,971 @@
+                        break;
+
+                    case "error":
+                        stderr.printf("%s\n", type_and_serialized[1]);
```
critical(), as above, save error, show to user

phako commented on src/faces/FacesTool.vala:
```diff
@@ -0,0 +1,971 @@
+
+                string[] name_and_value = piece.split("=");
+                if (name_and_value.length != 2) {
+                    stderr.printf("Wrong serialized line in face detection program output.");
```
critical()

phako reviewed (changes requested):
Please fix those, then we are good to go.

phako commented:
The very first task after the merge should be fixing the horrid user experience of editing the faces names. It is completely annoying that a small window with just an entry pops up

ricardofantin committed:
FacesTool.vala uses critical() and warning() instead of stderr

ricardofantin commented:
Everything corrected.

I looked how Digicam show faces, and soon I will make a better interface to define the person's name.

phako commented on src/faces/FacesTool.vala:
```diff
@@ -337,7 +337,7 @@ public class FacesTool : EditingTools.EditingTool {
             } catch (SpawnError e) {
-                stderr.printf("Error trying to spawn face detection program: %s\n", e.message);
+                critical("Error trying to spawn face detection program: %s\n", e.message);
                 assert_not_reached();
```
You missed one comment here. This is not a programming error, hence I would not assert.
I would rather remember that there was an error in the FaceDetectionJob, exit the function with "return", then check in the callback if the Job had an error and present it to the user with AppWindow.error_message

ricardofantin committed:
Included error messagem when cannot start the shotwell-facedetect

This commit change FacesTool.vala to show error message when the
program cannot call shotwell-facedetect to detect faces.

ricardofantin commented:
Good night.

I finally finish the error message when, for any reason, the program can't call shotwell-facedetect.
To see the error message compile and after delete facedetect/shotwell-facedetect .

The annoying floating entry is really bad.  I think something like digikam should be ok:
![digikam-editname](https://user-images.githubusercontent.com/5272036/39679000-803f8b5c-516d-11e8-951c-fdad279f164f.png)

The entry is with the main window, and not in a float. When the user use the detect faces it should be good change the label in facedetect window to entries as the user clicks.

Bye.

phako commented on src/faces/FacesTool.vala:
```diff
@@ -337,8 +337,10 @@ public class FacesTool : EditingTools.EditingTool {
-                assert_not_reached();
+                output = "Error trying to spawn face detection program: " + e.message + "\n";
+                stdout.printf("Erro: %s", output);
+                critical(output);
```
You reintroduced a printf here

phako commented on src/faces/FacesTool.vala:
```diff
@@ -950,8 +952,12 @@ public class FacesTool : EditingTools.EditingTool {
+            AppWindow.error_message(face_detection.output);
+            faces_tool_window.set_editing_phase(EditingPhase.DETECTING_FACES_FINISHED);
+		}else
+            pick_faces_from_autodetected();
```
Please fix the code style, that indentation looks off

phako commented on src/faces/FacesTool.vala:
```diff
@@ -337,8 +337,10 @@ public class FacesTool : EditingTools.EditingTool {
             } catch (SpawnError e) {
-                critical("Error trying to spawn face detection program: %s\n", e.message);
-                assert_not_reached();
+                output = "Error trying to spawn face detection program: " + e.message + "\n";
```
If this is shown to the user, it needs to be marked translatable. Why don't you save the SpawnError directly? Then you only have to do a != null check below.

ricardofantin commented:
Hello,
the last three errors were correct.

I can't wait to, after the merge, start asking to make the face-detection option to be default to true.

I have a deadline in next day 28, after that I'll have time to make the user interface changes.

Bye.

phako commented:
There are some minor style issues which I will fix while merging. I will merge this now.

For the GUI enhancements, you can use a PR on Shotwell's new GNOME Gitlab instance https://gitlab.gnome.org/GNOME/shotwell

phako commented:
Pushed to master
